### PR TITLE
I2C Wrapper + Controller

### PR DESCRIPTION
**Pull request by echavemann**
Mirror of the early work from Lab 6 + Adding I2C Controller
- prereq for #35

### CODE REVIEW

**Comment by echavemann** [comment]
TODO: include the i2c manager class in this so we don't have to deal with it in userland

File: external/nrf52x-base/sdk/nrf5_sdk_16.0.0/integration/nrfx/legacy/nrf_drv_twi.h [hunk 1]
```diff
@@ -161,9 +161,9 @@ typedef struct
  */
 #define NRF_DRV_TWI_DEFAULT_CONFIG                                               \
 {                                                                                \
-    .frequency          = (nrf_drv_twi_frequency_t)TWI_DEFAULT_CONFIG_FREQUENCY, \
     .scl                = 31,                                                    \
     .sda                = 31,                                                    \
+    .frequency          = (nrf_drv_twi_frequency_t)TWI_DEFAULT_CONFIG_FREQUENCY, \
```
Comment (echavemann) [inline external/nrf52x-base/sdk/nrf5_sdk_16.0.0/integration/nrfx/legacy/nrf_drv_twi.h:166]: yea the C++ requires that configs are declared in initialization order, I think the change is probably fine

File: src/hal/i2c_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#include "hal/i2c_controller.hpp"
+#include "hal/i2c_wrapper.hpp"
+#include "nrf_drv_twi.h"
+#include "microbit_v2.h"
+#include "nrf_twi.h"
+#include "nrf_twi_mngr.h"
+
+namespace edge::aidan
+{
+NRF_TWI_MNGR_DEF(twi_mngr, 1, 0);
```
Comment (echavemann) [inline src/hal/i2c_controller.cpp:10]: not sure where exactly is the best place to put this

File: external/nrf52x-base/sdk/nrf5_sdk_16.0.0/integration/nrfx/legacy/nrf_drv_twi.h (hunk 1 above)
Comment (stevenewald) [reply external/nrf52x-base/sdk/nrf5_sdk_16.0.0/integration/nrfx/legacy/nrf_drv_twi.h:166]: lmao

File: src/hal/i2c_controller.cpp (hunk 2 above)
Comment (stevenewald) [reply src/hal/i2c_controller.cpp:10]: looks fine. pls name constants in the future

File: include/hal/i2c_controller.hpp [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+namespace edge::aidan
+{
+
+class I2CController 
+{
+public:
+    ~I2CController();
```
Comment (stevenewald) [inline include/hal/i2c_controller.hpp:9]: I didn't see this defined. Just default it

File: src/hal/i2c_wrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    ret_code_t result = nrf_twi_mngr_perform(i2c_manager, NULL, write_transfer, 1, NULL);
+    if (result != NRF_SUCCESS) 
+    {
+        printf("I2C transaction failed! Error: %lX\n", result);
+    }
+}
+
+void lsm303agr_init()
+{
+    i2c_reg_write(LSM303AGR_ACC_ADDRESS, CTRL_REG5_A, 0x80);
```
Comment (stevenewald) [inline src/hal/i2c_wrapper.cpp:43]: ethan...

**Review by stevenewald — CHANGES_REQUESTED** [review]

File: src/hal/i2c_wrapper.cpp (hunk 4 above)
Comment (stevenewald) [reply src/hal/i2c_wrapper.cpp:43]: delete copied code

File: src/hal/i2c_controller.cpp (hunk 2 above)
Comment (echavemann) [reply src/hal/i2c_controller.cpp:10]: ill constexpr em

File: src/hal/i2c_wrapper.cpp (hunk 4 above)
Comment (echavemann) [reply src/hal/i2c_wrapper.cpp:43]: lol mb

File: src/hal/i2c_controller.cpp (hunk 2 above)
Comment (echavemann) [reply src/hal/i2c_controller.cpp:10]: macro doesnt like an int for id, not really torn up about fighting it

File: include/hal/i2c_controller.hpp (hunk 3 above)
Comment (echavemann) [reply include/hal/i2c_controller.hpp:9]: nice catch, changed

**Review by stevenewald — APPROVED** [review]
LGTM, change what u like im fine w it

**Pull request merged**